contrib/construct_dictionary.py

```diff
@@ -220,79 +221,98 @@ def main():
-    os.chdir(path)
+    src_include_dir = os.path.join(top_src_dir, "include")
+    build_include_dir = os.path.join(top_build_dir, "include")
 
     if options.dryrun:
         constants = sys.stdout
+        outpath = None
     else:
-        outpath = os.path.join(top, "dictionary.tmp")
+        outpath = os.path.join(build_include_dir, "dictionary.tmp")
         try:
             constants = open(outpath, "w+")
         except:
             print("DICTIONARY COULD NOT BE CONSTRUCTED")
-            os.chdir(save)
             return 1
 
     # write the header
-    constants.write("/*\n * This file is autogenerated by construct_dictionary.py.\n * Do not edit this file by hand.\n */\n\n")
-    constants.write("#include \"src/include/pmix_config.h\"\n")
-    constants.write("#include \"src/include/pmix_globals.h\"\n\n")
+    constants.write("""/*
+ * This file is autogenerated by construct_dictionary.py.
+ * Do not edit this file by hand.
+ */
+
+#include "src/include/pmix_config.h"
+#include "src/include/pmix_globals.h"
 
-    # write the starting line
-    constants.write("pmix_regattr_input_t dictionary[] = {\n")
+pmix_regattr_input_t dictionary[] = {
+""")
 
     # scan across the header files in the src directory
     # looking for constants and typedefs
-    rc = harvest_constants(options, "pmix_common.h", constants)
+
+    # pmix_common.h is in the build tree
+    rc = harvest_constants(options,
+                           os.path.join(build_include_dir, "pmix_common.h"),
+                           constants)
     if 0 != rc:
         constants.close()
-        os.remove(outpath)
+        if outpath:
+            os.remove(outpath)
         print("HARVEST PMIX_COMMON FAILED - DICTIONARY COULD NOT BE CONSTRUCTED")
-        os.chdir(save)
         return 1
     constants.write(",\n")
-    rc = harvest_constants(options, "pmix_deprecated.h", constants)
+
+    # pmix_deprecated.h is in the source tree
+    rc = harvest_constants(options,
+                           os.path.join(src_include_dir, "pmix_deprecated.h"),
+                           constants)
     if 0 != rc:
         constants.close()
-        os.remove(outpath)
+        if outpath:
+            os.remove(outpath)
         print("HARVEST DEPRECATED FAILED - DICTIONARY COULD NOT BE CONSTRUCTED")
-        os.chdir(save)
         return 1
 
     # mark the end of the array
-    constants.write(",\n    {.name = \"\"}\n")
-    # write the closure
-    constants.write("\n};\n")
+    constants.write(""",
+    {.name = ""}
+};
+""")
 
     # transfer the results
-    target = os.path.join(top, "src", "include", "dictionary.h")
-    try:
-        os.replace(outpath, target)
-    except:
-        print("DICTIONARY COULD NOT BE CONSTRUCTED")
-        os.chdir(save)
-        return 1
-
-    # return to original directory
-    os.chdir(save)
+    # Write to the source tree
+    if outpath:
+        try:
+            target = os.path.join(src_include_dir, "dictionary.h")
```

Comment:
@rhc54 This seems wrong, for two reasons:

1. We should not be writing to the source tree.
2. Writing to the build tree after the build seems like a contradiction.  This script is invoked during `make install`, after PMIx is built... but then we go back and re-write `src/include/dictionary.h`.  Shouldn't `dictionary.h` be written before the build?  And if `dictionary.h` is written at build time, shouldn't it be in the build tree?  And if it should be in the build tree, should it not be in the source tree?